File: src/main/java/cube8540/oauth/authentication/credentials/resource/infra/SecuredResourceExceptionTranslator.java
package cube8540.oauth.authentication.credentials.resource.infra;

import cube8540.oauth.authentication.credentials.resource.domain.exception.ResourceInvalidException;
import cube8540.oauth.authentication.credentials.resource.domain.exception.ResourceNotFoundException;
import cube8540.oauth.authentication.credentials.resource.domain.exception.ResourceRegisterException;
import cube8540.oauth.authentication.error.ExceptionTranslator;
import cube8540.oauth.authentication.error.message.ErrorMessage;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

@Slf4j
public class SecuredResourceExceptionTranslator implements ExceptionTranslator<ErrorMessage<Object>> {
    @Override
    public ResponseEntity<ErrorMessage<Object>> translate(Exception exception) {
        if (exception instanceof ResourceNotFoundException) {
            ResourceNotFoundException e = ((ResourceNotFoundException) exception);
            return new ResponseEntity<>(ErrorMessage.instance(e.getCode(), e.getDescription()), HttpStatus.NOT_FOUND);
        } else if (exception instanceof ResourceRegisterException) {
            ResourceRegisterException e = ((ResourceRegisterException) exception);
            return new ResponseEntity<>(ErrorMessage.instance(e.getCode(), e.getDescription()), HttpStatus.BAD_REQUEST);
        } else if (exception instanceof ResourceInvalidException) {
            ResourceInvalidException e = ((ResourceInvalidException) exception);
            return new ResponseEntity<>(ErrorMessage.instance(e.getCode(), e.getErrors().toArray()), HttpStatus.BAD_REQUEST);
        } else {
            log.error("Handle exception {}, {}", exception.getClass(), exception.getMessage());
            return response(HttpStatus.INTERNAL_SERVER_ERROR, ErrorMessage.UNKNOWN_SERVER_ERROR);
        }
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/resource/infra/SecuredResourceExceptionTranslator.kt
package cube8540.oauth.authentication.resource.infra

import cube8540.oauth.authentication.resource.domain.ResourceInvalidException
import cube8540.oauth.authentication.resource.domain.ResourceNotFoundException
import cube8540.oauth.authentication.resource.domain.ResourceRegisterException
import cube8540.oauth.authentication.error.ExceptionTranslator
import cube8540.oauth.authentication.error.message.ErrorMessage
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Component

@Component
class SecuredResourceExceptionTranslator: ExceptionTranslator<ErrorMessage<Any>> {

    private val logger = LoggerFactory.getLogger(this::class.java)

    override fun translate(exception: Exception): ResponseEntity<ErrorMessage<Any>> = when(exception) {
        is ResourceNotFoundException -> {
            ResponseEntity(ErrorMessage.instance(exception.code, exception.message), HttpStatus.NOT_FOUND)
        }
        is ResourceRegisterException -> {
            ResponseEntity(ErrorMessage.instance(exception.code, exception.message), HttpStatus.BAD_REQUEST)
        }
        is ResourceInvalidException -> {
            ResponseEntity(ErrorMessage.instance(exception.code, exception.errors), HttpStatus.BAD_REQUEST)
        }
        else -> {
            logger.error("Handle exception {} {}", exception.javaClass, exception.message)
            response(HttpStatus.INTERNAL_SERVER_ERROR, ErrorMessage.UNKNOWN_SERVER_ERROR)
        }
    }
}